Network/Socket/Types.hsc

```diff
@@ -108,12 +116,19 @@ fdSocket (Socket ref _) = readIORef ref
 
 -- | Creating a socket from a file descriptor.
 mkSocket :: CInt -> IO Socket
-mkSocket fd = do
+mkSocket !fd = do
     ref <- newIORef fd
     let s = Socket ref fd
-    void $ mkWeakIORef ref $ close s
+    void $ mkWeakFromIORef ref () $ close s
     return s
 
+-- | Make a weak pointer from an 'IORef' key to an arbitrary
+-- value. This is similar to 'Data.IORef.mkWeakIORef', but we use it
+-- to avoid keeping the useless 'IORef' wrapper alive.
+mkWeakFromIORef :: IORef a -> b -> IO () -> IO (Weak b)
+mkWeakFromIORef (IORef (STRef r##)) b (IO finalizer) = IO $ \s ->
```

Comment:
I'd also love to see this function move to its own internal module.

Reply:
Agreed. I'll move it.